enterprise/server/usage/usage.go

```diff
@@ -83,41 +128,235 @@ func (ut *tracker) Increment(ctx context.Context, uc *tables.UsageCounts) error
+				if err := ut.FlushToDB(ctx); err != nil {
+					log.Errorf("Failed to flush usage data to DB: %s", err)
+				}
+			case <-ut.stopFlush:
+				return
+			}
+		}
+	}()
+}
+
+// StopDBFlush cancels the goroutine started by StartDBFlush.
+func (ut *tracker) StopDBFlush() {
+	ut.stopFlush <- struct{}{}
+}
+
+// FlushToDB flushes usage metrics from any finalized collection periods to the
+// DB.
+//
+// Public for testing only; the server should call StartDBFlush to periodically
+// flush usage.
+func (ut *tracker) FlushToDB(ctx context.Context) error {
+	// redsync.ErrFailed can be safely ignored; it just means that another app
+	// is already trying to flush.
+	if err := ut.flushToDB(ctx); err != nil && err != redsync.ErrFailed {
+		return err
+	}
+	return nil
+}
+
+func (ut *tracker) flushToDB(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, flushTimeout)
+	defer cancel()
+
+	// Grab lock. This will immediately return `redsync.ErrFailed` if another
+	// client already holds the lock.
+	if err := ut.flushMutex.LockContext(ctx); err != nil {
+		return err
+	}
+	defer func() {
+		ctx, cancel := background.ExtendContextForFinalization(ctx, redisUsageLockFinalizationTimeout)
+		defer cancel()
+		if ok, err := ut.flushMutex.UnlockContext(ctx); !ok || err != nil {
+			log.Warningf("Failed to unlock Redis lock: ok=%v, err=%s", ok, err)
+		}
+	}()
+
+	// Loop through collection periods starting from the oldest collection period
+	// that may exist in Redis (based on key expiration time) and looping up until
+	// we hit a collection period which is not yet "settled".
+
+	for c := ut.oldestWritableCollectionPeriod(); ut.isSettled(c); c = c.Next() {
+		// Read groups
+		gk := groupsRedisKey(c)
+		groupIDs, err := ut.rdb.SMembers(ctx, gk).Result()
+		if err != nil {
+			return err
+		}
+
+		for _, groupID := range groupIDs {
+			// Read usage counts from Redis
+			ck := countsRedisKey(groupID, c)
+			h, err := ut.rdb.HGetAll(ctx, ck).Result()
+			if err != nil {
+				return err
+			}
+			counts, err := stringMapToCounts(h)
+			if err != nil {
+				return err
+			}
+			// Update counts in the DB
+			if err := ut.flushCounts(ctx, groupID, c, counts); err != nil {
+				return err
+			}
+			// Clean up the counts from Redis
+			if _, err := ut.rdb.Del(ctx, ck).Result(); err != nil {
+				return err
+			}
+		}
+
+		// Delete the Redis data for the groups.
+		if _, err := ut.rdb.Del(ctx, gk).Result(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ut *tracker) flushCounts(ctx context.Context, groupID string, c collectionPeriod, counts *tables.UsageCounts) error {
+	pk := &tables.Usage{
+		GroupID:         groupID,
+		PeriodStartUsec: timeutil.ToUsec(c.UsagePeriod().Start()),
+	}
+	dbh := ut.env.GetDBHandle()
+	return dbh.Transaction(ctx, func(tx *db.DB) error {
+		// Create a row for the corresponding usage period if one doesn't already
+		// exist
+		err := tx.Clauses(clause.OnConflict{DoNothing: true}).Create(pk).Error
```

Comment:
Could you do INSERT ON DUPLICATE KEY IGNORE or something instead of relying on special gorm behavior here? I think it would be clearer to the reader and also not depend on gorm intricacies.